org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java

```diff
@@ -0,0 +1,157 @@
+import org.eclipse.lsp4j.RegistrationParams;
+import org.eclipse.lsp4j.Unregistration;
+import org.eclipse.lsp4j.UnregistrationParams;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4xml.XMLTextDocumentService;
+import org.eclipse.lsp4xml.settings.XMLFormattingOptions;
+/**
+ * ClientPreferences
+ */
+public class XMLCapabilityManager {
+
+	//ClientCapabilityPreferences clientCapabilityPreferences;
+	ClientCapabilitiesWrapper clientWrapper;
+	private Set<String> registeredCapabilities = new HashSet<>(3);
+	private LanguageClient languageClient;
+	private XMLTextDocumentService textDocumentService;
+
+	public XMLCapabilityManager(LanguageClient languageClient) {
+		this.languageClient = languageClient;
+	}
+
+	public void setTextDocumentService(XMLTextDocumentService xmlTextDocumentService) {
+		this.textDocumentService = xmlTextDocumentService;
+	}
+
+	public XMLTextDocumentService getTextDocumentService() {
+		return this.textDocumentService;
+	}
+
+	// public void setClientCapabilityPreferences(ClientCapabilityPreferences clientCapabilityPreferences) {
+	// 	this.clientCapabilityPreferences = clientCapabilityPreferences;
+	// }
+
+	// public ClientCapabilityPreferences getClientCapabilityPreferences() {
+	// 	return this.clientCapabilityPreferences;
+	// }
+
+	public void updateClientCapabilitiesWrapper(ClientCapabilities clientCapabilities) {
+		
+		this.clientWrapper = new ClientCapabilitiesWrapper(clientCapabilities);
+	}
+
+	public ClientCapabilitiesWrapper getClientCapabilitiesWrapper() {
+		if(this.clientWrapper == null) {
+			this.clientWrapper = new ClientCapabilitiesWrapper();
+		}
+		return this.clientWrapper;
+	}
+
+	/**
+	 * 
+	 */
+	private void toggleCapability(boolean enabled, String id, String capability, Object options) {
+		if (enabled) {
+			registerCapability(id, capability, options);
+		} else {
+			unregisterCapability(id, capability);
+		}
+	}
+
+	public void unregisterCapability(String id, String method) {
+		if (registeredCapabilities.remove(id)) {
+			Unregistration unregistration = new Unregistration(id, method);
+			UnregistrationParams unregistrationParams = new UnregistrationParams(
+					Collections.singletonList(unregistration));
+			languageClient.unregisterCapability(unregistrationParams);
+		}
+	}
+
+	public void registerCapability(String id, String method) {
+		registerCapability(id, method, null);
+	}
+
+	public void registerCapability(String id, String method, Object options) {
+		if (registeredCapabilities.add(id)) {
+			Registration registration = new Registration(id, method, options);
+			RegistrationParams registrationParams = new RegistrationParams(Collections.singletonList(registration));
+			languageClient.registerCapability(registrationParams);
+		}
+	}
+
+	public void initializeCapabilities() {
+		if (this.getClientCapabilitiesWrapper().isCodeActionDynamicRegistered()) {
+			registerCapability(Preferences.CODE_ACTION_ID, Preferences.TEXT_DOCUMENT_CODE_ACTION, null);
```

Comment:
import static org.eclipse.lsp4xml.settings.capabilities.ServerCapabilitiesConstants.*; so you don't have to add the class prefix before each constant